kuebiko/pipeline/nspl/flow.py

```diff
@@ -0,0 +1,215 @@
+    pcd_laua: Dict[postcode, laua_code]
+    pcd_latlong: Dict[postcode, Dict[Literal["lat", "long"], float]]
+
+    @step
+    def start(self):
+        """Get dynamically rendered download link."""
+        from utils import (
+            chrome_driver,
+            find_download_url,
+        )
+
+        self.geoportal_url = f"{GEOPORTAL_URL_PREFIX}/{self.geoportal_dataset}"
+        with chrome_driver() as driver:
+            self.download_url = find_download_url(driver, self.geoportal_url)
+
+        self.next(self.get_data)
+
+    @step
+    def get_data(self):
+        """Download zipped NSPL collection, extracting main lookup & LAUA names."""
+        import logging
+
+        import requests_cache
+
+        from utils import (
+            download_zip,
+            extract_laua_year,
+            filter_nspl_data,
+            read_nspl_data,
+            read_laua_names,
+        )
+
+        ## **If a flow is run with `--production` (`current.is_production` is
+        ## `True`) then it should NOT run in test mode.**
+        if self.test_mode and not current.is_production:
+            nrows = 1_000
+            logging.warning(f"TEST MODE: Constraining to first {nrows} rows!")
+        else:
+            nrows = None
+
+        ## Downloading a large zip each time we iterate is inefficient.<br>
+        ## With other expensive tasks we can split them into different
+        ## metaflow steps and `resume` part-way through our flow; however
+        ## if we download a zip in one step and process it in another then
+        ## this requires that the two steps run on the same machine. One of
+        ## the big benefits of metaflow is how easy it is to run a pipeline in
+        ## a distributed manner.
+        ## If you knew the flow would only ever be run locally with all steps
+        ## happening on the same machine then it would _probably_ be ok to do
+        ## the above. But we can have the best of both worlds by using
+        ## `requests_cache` to cache requests (i.e. downloading the zip) for us.
+        ## We can also make sure that if the `--production` flag is used then
+        ## the file is re-downloaded rather than risking an invalid cache/local copy.
+        if not current.is_production:
+            requests_cache.install_cache("nspl_zip_cache")
+        with download_zip(self.download_url) as zipfile:
+            # Load main postcode lookup
+            self.nspl_data = read_nspl_data(zipfile, nrows).pipe(filter_nspl_data)
+
+            # LAUA lookup from codes to names
+            laua_names_tmp = read_laua_names(zipfile)
+            self.laua_year = extract_laua_year(laua_names_tmp)
+            ## We could have extracted year in `read_laua_names` but we keep to
+            ## the single responsibility principle.
+            self.laua_names = laua_names_tmp.to_dict()
+
+        self.next(self.data_quality)
+
+    @step
+    def data_quality(self):
+        """Data quality checks."""
```

Comment:
@bishax This is a great step - would it be even better to package each of the checks as a separate function? More generally, should we think about data tests similarly to function unit tests?